File: java/src/main/java/parser/ast/ASTNode.java
package parser.ast;

import lexer.Token;
import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;

public abstract class ASTNode {
    /*树*/
    protected ArrayList<ASTNode> children =new ArrayList<>();
    protected ASTNode parent;

    /*关键信息*/
    protected Token lexeme;
    protected String label;
    protected ASTNodeTypes type;


    public ASTNode(ASTNode _parent){
        this.parent = _parent;
    }
    public ASTNode(ASTNode _parent,ASTNodeTypes _type, String _label){
        this.parent=_parent;
        this.type=_type;
        this.label=_label;
    }

    public ASTNode getChild(int index){
        return this.children.get(index);
    }

    public void addChild(ASTNode node){
        children.add(node);
    }

    public Token getLexeme(){
        return lexeme;
    }

    public List<ASTNode> getChildren(){
        return children;
    }

    public void setLexeme(Token token){
        this.lexeme=token;
    }
    public void setLabel(String s){
        this.label=s;
    }

    public  void setType(ASTNodeTypes type){
        this.type=type;
    }

    public void print(int indent){
        System.out.println(StringUtils.leftPad(" ",indent*2)+label);
        for (ASTNode child:children){
            child.print(indent+1);
        }
    }


}

File: js/parser/ast/ASTNode.js
class ASTNode{
    constructor(_parent,_type=null,_label = null) {
        // 数结构
        this.children =[];
        this.parent = _parent;

        // 关键信息
        this.lexeme = null
        this.type = _type
        this.label = _label
    }

    getChild(index){
        return this.children[index]
    }

    addChild(node){
        this.children.push(node)
    }

    getLexeme(){
        return this.lexeme
    }

    getChildren(){
        return this.children
    }

    print(indent=0){
        console.log(`${"".padStart(indent*2," ")}${this.label}`)
        this.children.forEach((child)=>{
            child.print(indent+1)
        })


    }

}

module.exports=ASTNode